### PR TITLE
[th/wireguard-pt3]

### PR DESCRIPTION
**Pull request by thom311**
A major part for WireGuard support.

Still missing:

- nmcli support (but there is an example script that can be used instead)
- support for preshared-key (and requesting secrets)
- more testing
- see `TODO.txt` file

But I think it's at a point where it could be merged. We can make it perfect afterwards.

### CODE REVIEW

**Comment by thom311** [comment]
I opened https://github.com/NetworkManager/NetworkManager/pull/282, which contains the early part of this branch. It is mostly preparation work, let's split this pull request in smaller chunks. If you care, review https://github.com/NetworkManager/NetworkManager/pull/282 first.

File: libnm-core/nm-setting-wireguard.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+};
+
+G_DEFINE_TYPE (NMSettingWireGuard, nm_setting_wireguard, NM_TYPE_SETTING)
+
+#define NM_SETTING_WIREGUARD_GET_PRIVATE(self) _NM_GET_PRIVATE (self, NMSettingWireGuard, NM_IS_SETTING_WIREGUARD, NMSetting)
+
+/*****************************************************************************/
+
+const char *
+nm_setting_wireguard_get_private_key (NMSettingWireGuard *self)
```
Comment (bengal) [inline libnm-core/nm-setting-wireguard.c]: Missing documentation comment

File: libnm-core/nm-setting-wireguard.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+	object_class->finalize     = finalize;
+
+	setting_class->verify         = verify;
+	setting_class->verify_secrets = verify_secrets;
+	setting_class->need_secrets   = need_secrets;
+
+	/**
+	 * NMSettingWireGuard:private-key:
+	 *
+	 * The 128 bit private-key in base64 encoding.
```
Comment (bengal) [inline libnm-core/nm-setting-wireguard.c]: Isn't it 256 bit?

File: libnm-core/nm-setting-wireguard.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+	                         NULL,
+	                           G_PARAM_READWRITE
+	                         | NM_SETTING_PARAM_SECRET
+	                         | G_PARAM_STATIC_STRINGS);
+
+	/**
+	 * NMSettingWireGuard:private-key-flags:
+	 *
+	 * Flags indicating how to handle the #NMSettingWirelessSecurity:psk
+	 * property.
```
Comment (bengal) [inline libnm-core/nm-setting-wireguard.c]: Wrong property name.

File: src/devices/nm-device-wireguard.c [hunk 4]
```diff
@@ -114,36 +120,299 @@ link_changed (NMDevice *device,
...
+{
+	const char *iface = nm_device_get_iface (device);
+	int r;
+
+	g_return_val_if_fail (iface, FALSE);
+
+	r = nm_platform_link_wireguard_add (nm_device_get_platform (device), iface, out_plink);
+	if (r < 0) {
+		g_set_error (error, NM_DEVICE_ERROR, NM_DEVICE_ERROR_CREATION_FAILED,
+		             "Failed to create bond interface '%s' for '%s': %s",
```
Comment (bengal) [inline src/devices/nm-device-wireguard.c]: bond -> wireguard

**Comment by thom311** [comment]
addressed comments so far and repushed.

**Comment by thom311** [comment]
closing for the moment, to wait for https://github.com/NetworkManager/NetworkManager/pull/288 and https://github.com/NetworkManager/NetworkManager/pull/291 . Will be reopened later.

**Pull request closed without merging**